Saved exports progress bar

esoergel opened the pull request:
Adds a progress bar to the saved exports page:
![image](https://user-images.githubusercontent.com/2367539/41489228-3a041c38-70bd-11e8-85c8-f521c2fef9fa.png)


:tropical_fish: 

@czue

esoergel committed:
Move export tasks into exports module

esoergel committed:
Track progress of saved export rebuilds

esoergel committed:
Add endpoint to get task status

esoergel committed:
Make functional progress bar!!

esoergel committed:
Minor simplification

fmt_emailed_export_data is only called once, and not all the arguments are
provided.  Just combine the two functions.  I also noticed that
_fmt_emailed_export_fileData is only called if has_file, so there's no need to
pass in has_file

esoergel committed:
Show progress bar on page load if active task

esoergel committed:
Notify on success

esoergel committed:
Start progress bar also on filter change

esoergel committed:
Route all saved export task rebuilds through one place

esoergel committed:
Use only two queues for saved exports

czue commented on corehq/apps/export/export.py:
```diff
@@ -456,11 +456,11 @@ def _get_base_query(export_instance):
     Rebuild the given daily saved ExportInstance
     """
-    filters = filters or export_instance.get_filters()
+    filters = export_instance.get_filters()
```
nice cleanup in this commit. it's a bit hard to wrap my head around this, but do you know why this logic used to be here and why filters were previously passed in separately?

czue replied:
I think maybe the answer is that filters was never actually passed in and the reason must be "some legacy thing that no longer matters"

esoergel replied:
> "some legacy thing that no longer matters"

That describes a lot of this code.  I suspect that at one point it was passed in (prior to [this commit](https://github.com/dimagi/commcare-hq/commit/54ca07a4c5c24a6c7935d4f2bf7e630d8f33802b) perhaps), then that was probably phased out, but the optional arg was never removed.

czue reviewed (approved):
was a tiny bit hard for me to follow some of the task refactors, but overall this looks great and like a nice cleanup. 

excited for this feature!

czue commented:
fyi front end tests are failing

```
  1) ListExportsController Unit Tests Actions selectAll():
     TypeError: 'undefined' is not an object (evaluating 'exp.emailedExport.taskStatus.inProgress')
      at http://localhost:8000/static/export/js/list_exports.ng.js:49
      at http://localhost:8000/static/underscore/underscore-min.js:5
      at http://localhost:8000/static/export/js/list_exports.ng.js:52
      at http://localhost:8000/static/angular/angular.min.js:87
      at http://localhost:8000/static/angular/angular.min.js:118
      at http://localhost:8000/static/angular/angular.min.js:133
      at http://localhost:8000/static/angular/angular.min.js:130
      at http://localhost:8000/static/angular/angular.min.js:133
      at g (http://localhost:8000/static/angular/angular.min.js:87)
      at handleResponse (http://localhost:8000/static/angular-mocks/angular-mocks.js:1158)
      at http://localhost:8000/static/angular-mocks/angular-mocks.js:1516
      at http://localhost:8000/static/export/spec/ListExportsController.spec.mo.js:133
      at callFn (http://localhost:8000/static/mocha/mocha.js:4202)
      at http://localhost:8000/static/mocha/mocha.js:4195
      at http://localhost:8000/static/mocha/mocha.js:4661
      at http://localhost:8000/static/mocha/mocha.js:4790
      at next (http://localhost:8000/static/mocha/mocha.js:4581)
      at http://localhost:8000/static/mocha/mocha.js:4591
      at next (http://localhost:8000/static/mocha/mocha.js:4523)
      at http://localhost:8000/static/mocha/mocha.js:4554
      at done (http://localhost:8000/static/mocha/mocha.js:4163)
      at callFn (http://localhost:8000/static/mocha/mocha.js:4217)
      at http://localhost:8000/static/mocha/mocha.js:4195
      at next (http://localhost:8000/static/mocha/mocha.js:4555)
      at http://localhost:8000/static/mocha/mocha.js:4559
      at timeslice (http://localhost:8000/static/mocha/mocha.js:12326)
  2) ListExportsController Unit Tests Actions "after each" hook for "selectAll()":
     Error: [$rootScope:inprog] http://errors.angularjs.org/1.4.4/$rootScope/inprog?p0=%24digest
      at q (http://localhost:8000/static/angular/angular.min.js:125)
      at http://localhost:8000/static/angular/angular.min.js:129
      at http://localhost:8000/static/angular-mocks/angular-mocks.js:1538
      at http://localhost:8000/static/export/spec/ListExportsController.spec.mo.js:126
      at callFn (http://localhost:8000/static/mocha/mocha.js:4202)
      at http://localhost:8000/static/mocha/mocha.js:4195
      at next (http://localhost:8000/static/mocha/mocha.js:4555)
      at http://localhost:8000/static/mocha/mocha.js:4559
      at timeslice (http://localhost:8000/static/mocha/mocha.js:12326)
```

esoergel commented on corehq/apps/reports/tasks.py:
```diff
@@ -145,40 +142,6 @@ def monthly_reports():
-
-
-@task(queue='background_queue', ignore_result=True)
-def rebuild_export_task(groupexport_id, index, last_access_cutoff=None, filter=None):
```
Note that this is `corehq.apps.reports.tasks.rebuild_export_task`, which is _not_ the `corehq.apps.export.tasks.rebuild_export_task` called above.  It doesn't appear to be called anywhere, so this commit drops it.

esoergel committed:
Namespace taskStatus response

esoergel committed:
Update js tests

esoergel committed:
Move deduplication into rebuild_saved_export

Previously, the first queued export task would be allowed to succeed,
and subsequent tasks would queue but noop.  This version looks for
pending or running tasks and if they exist, doesn't queue a new task,
unless the task was queued manually and the pending task hasn't started
yet.

esoergel committed:
lint [ci skip]

esoergel commented:
This should be ready to go now.  Might try to throw it on staging for some QA.

Here's the (currently pending) build: https://travis-ci.org/dimagi/commcare-hq/builds/393853283

czue commented on corehq/apps/export/tasks.py:
```diff
@@ -81,14 +81,9 @@ def populate_export_download_task(export_instances, filters, download_id, filena
 def _start_export_task(export_instance_id, last_access_cutoff):
-    keys = ['rebuild_export_task_%s' % export_instance_id]
-    timeout = 48 * 3600  # long enough to make sure this doesn't get called while another one is running
-    with CriticalSection(keys, timeout=timeout, block=False) as locked_section:
```
A little hard to wrap my head around this, but it seems like you moved a locking/concurrency check from inside the task to the caller of the task, yeah? Just thinking out loud I could imagine that could cause problems in rare scenarios - e.g. you check the task status from two threads at the same time, both see there is nothing there, then both start the task at the same instant resulting in the check not working and the task running in parallel. Realize it's a bit of an edge case, but seems less safe than doing it in the task. Granted, I don't really understand why this was in a `CriticalSection` in the first place or what problems arise if that ever trips. 

Does this commit fix a bug? Or just result in fewer noops in the queue? Just wondering if we can do the check outside the task but keep the `CriticalSection` logic in place for edge cases.

esoergel replied:
Saved exports have a single place for a payload, so if two tasks run at the same time, one will finish, be saved, and then the second one will finish and overwrite the result of the first task.  These are non-trivially expensive to compute, which I believe is why this is locked.

That was added in https://github.com/dimagi/commcare-hq/pull/17848.  (@snopoke you may be interested in this change.)

The impetus for this last commit is that this progress bar operates on the task itself.  If a task is queued, it then becomes _the_ task for that export instance, and we lose the pointer to any previous task.  Basically, if a task is in progress, and another one is queued, it will noop, and the user will see the (senseless) progress bar from the second task.  To prevent this, we need to know whether the task will conflict before calling it.

I also considered just checking the lock before calling the task, but that wouldn't cover tasks that are queued but not yet started.  It wouldn't be difficult to add some locking back in, for instance in the code that queues a task, to be sure that's never executed in parallel.  No strong feelings there.  What do you think?

czue replied:
Thanks for the explanation. It sounds like the failure mode is basically "do some work twice" which, under the extremely rare edge case I pointed out, seems totally fine. So I think your change makes sense and no need for it to get any more complicated.

czue reviewed (commented):
Had one question / thought before merge. Maybe just helping me understand the last commit.
